File: PDOO-Civitas/CivitasR/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require 'io/console'

module Civitas

  class Vista_textual
    
    @separador = "================================= COMENZAMOS =================================="

    def mostrar_estado(estado)
      puts estado
    end
    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    def salirCarcel()
      lista_salir_carcel = [Salidas_carcel::PAGANDO,Salidas_carcel::TIRANDO]
    
      opcion = menu("Elige la forma para intentar salir de la carcel", ["Pagando","Tirando el dado"])
    
      return lista_salir_carcel[opcion]
    end
    
    def comprar
      lista_respuestas = [Respuestas::SI,Respuestas::NO]
      
      opcion = menu("¿Desea comprar la calle a la que se ha llegado?", ["SI","NO"])
    
      return lista_respuestas[opcion]
    end

    def gestionar
      
      opcion = menu("Indique el número de gestión inmobiliaria elegida",
      ["Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar"])
      
      ja = @juegoModel.get_jugador_actual()
      
      propiedades = ja.propiedades
      nombre_propiedades = Array.new
      
      var = 0
      
      while(var < propiedades.length())
        
        nombre_propiedades.push(propiedades.at(var).nombre)                
        var = var+1
      end
      
      if(opcion != 5)
        
        ip = menu("Indique el número de la propiedad a gestionar", nombre_propiedades)
      end
      
      @ipropiedad = ip
      @igestion = opcion
    end

    def getGestion
      return @igestion
    end

    def getPropiedad
      return @ipropiedad
    end

    def mostrarSiguienteOperacion(operacion)

      puts "\nSiguiente operación: " + operacion.to_s      
    end

    def mostrarEventos
      
      while(Diario.instance.eventos_pendientes())
        
        puts "\nEvento pendiente: " + Diario.instance.leer_evento()
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      
      puts "\nJugador Actual: " + @juegoModel.get_jugador_actual().to_s
      puts @separador
      puts "\nCasilla Actual: " + @juegoModel.get_casilla_actual().to_s
      puts @separador
      puts @separador
      
    end

    
  end

end

File: PDOO-Civitas/CivitasJ/src/GUI/VistaTextual.java
package GUI;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.Operaciones_juego;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.CivitasJuego;
import civitas.GestionesInmobiliarias;
import civitas.Jugador;
import civitas.SalidasCarcel;
import civitas.TituloPropiedad;

public class VistaTextual{
  
  CivitasJuego juegoModel; 
  private int iGestion=-1;
  private int iPropiedad=-1;
  private static String separador = "===================================================================";
  
  private Scanner in;
  
  public VistaTextual() {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar(){
  
    int opcion = menu ("¿Desea comprar la calle a la que se ha llegado?",
      new ArrayList<> (Arrays.asList("SI","NO")));
    
    return (Respuestas.values()[opcion]);
  }

  void gestionar(){
  
    int opcion = menu ("Indique el número de gestión inmobiliaria elegida",
      new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar")));
    
    Jugador ja = juegoModel.getJugadorActual();
    
    ArrayList<TituloPropiedad> propiedades = ja.getPropiedades();
    ArrayList<String> nombrePropiedades = new ArrayList<String>(); 
    
    for(int i = 0; i < propiedades.size(); i++){
        
        nombrePropiedades.add(propiedades.get(i).getNombre());
    }
    
    int ip = -1;
    
    if(opcion != 5){
    
        ip = menu ("Indique el número de la propiedad a gestionar",
      nombrePropiedades);
    }
    
    this.iPropiedad = ip;
    this.iGestion = opcion;
  }
  
  public int getGestion(){
  
      return this.iGestion;
  }
  
  public int getPropiedad(){
  
      return this.iPropiedad;
  }
    

  void mostrarSiguienteOperacion(Operaciones_juego operacion) {
  
      System.out.println("\nSiguiente operación: " + operacion.name());
  }


  void mostrarEventos(){
  
      while(Diario.getInstance().eventosPendientes()){
      
          System.out.println("\nEvento pendiente: " + Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  void actualizarVista(){
  
      System.out.println("\nJugador Actual: " + juegoModel.getJugadorActual().toString());
      System.out.println(separador);
      System.out.println("\nCasilla Actual: " + juegoModel.getCasillaActual().toString());
      System.out.println(separador);
      System.out.println(separador);
  } 
}
